devito/passes/iet/engine.py

```diff
@@ -221,72 +221,110 @@ def abstract_efunc(efunc):
     Abstract `efunc` applying a set of rules:
 
         * The `efunc` names becomes "foo".
-        * Any concrete AbstractFunction gets replaced with a "more abstract" object:
-            - DiscreteFunctions become AliasFunctions with name "f0", "f1", ...
-            - Arrays remain Arrays but are renamed as "a0", "a1", ...
-        * Objects remain Objects but are renamed as "o0", "o1", ...
+        * Symbolic objects get replaced with "more abstract" objects:
+            - DiscreteFunctions are renamed as "f0", "f1", ...
+            - Arrays are renamed as "a0", "a1", ...
+            - Objects are renamed as "o0", "o1", ...
     """
-    mapper = {}
-    counter = Counter()
-    for i in FindSymbols().visit(efunc):
-        if isinstance(i, AliasFunction):
-            base = i.name
-
-        elif i.is_DiscreteFunction:
-            base = 'f'
-
-            rkwargs = set(i.__rkwargs__) - {'initializer'}
-            kwargs = {k: getattr(i, k, None) for k in rkwargs}
-            kwargs['name'] = '%s%d' % (base, counter[base])
-            v = AliasFunction(aliased=i, **kwargs)
-
-            mapper.update({
-                i: v,
-                i.indexed: v.indexed,
-                i.dmap: v.dmap,
-                i._C_symbol: v._C_symbol,
-            })
-
-        elif i.is_Array:
-            # Use special names for special objects
-            if isinstance(i, Lock):
-                base = 'lock'
-            else:
-                base = 'a'
-
-            kwargs = {k: getattr(i, k, None) for k in i.__rkwargs__}
-            kwargs['name'] = '%s%d' % (base, counter[base])
-            v = type(i).__base__(**kwargs)
-
-            mapper.update({
-                i: v,
-                i.indexed: v.indexed,
-                i._C_symbol: v._C_symbol,
-            })
-
-        elif isinstance(i, CompositeObject):
-            base = 'o'
-
-            args = [getattr(i, k, None) for k in i.__rargs__]
-            args[0] = '%s%d' % (base, counter[base])
-            kwargs = {k: getattr(i, k, None) for k in i.__rkwargs__}
-            v = i.func(*args, **kwargs)
-
-            mapper[i] = v
+    functions = FindSymbols().visit(efunc)
 
-        else:
-            base = i.name
+    # Precedence rules make it possible to reconstruct objects that depend on
+    # higher priority objects
+    priority = [DiscreteFunction]
 
-            v = i
+    def key(i):
+        for p, cls in enumerate(priority, 1):
```

Comment:
Wouldn't it be better with priority as a Dict so that we can easily add new ones with explicit priorities?

Reply:
sounds good, changing